src/update-mp-v1-ids-to-v2.ts

```diff
@@ -0,0 +1,174 @@
+} from "graphql"
+import prettier from "prettier"
+import { ASTNode } from "recast"
+import { schema as promisedSchema } from "./helpers/getSchema"
+import { fragmentOnNonCompositeErrorMessage } from "graphql/validation/rules/FragmentsOnCompositeTypes"
+
+/**
+ * Follows the selection chain in a GQL operation and returns the _parent_ type
+ * from the related schema
+ */
+const walkSelections = (
+  selections: Array<string | number> | null,
+  operationChain: DefinitionNode,
+  schemaPosition: any
+): DefinitionNode => {
+  if (selections !== null && selections.length > 2) {
+    let selection = selections.shift()!
+    let definition = operationChain[selection] as DefinitionNode
+    let newSchemaPosition = schemaPosition
+    if (typeof selection === "number") {
+      newSchemaPosition = schemaPosition.fields.find(
+        field => definition.name.value === field.name.value
+      )
+    }
+    return walkSelections(selections, definition, newSchemaPosition)
+  }
+  return schemaPosition
+}
+
+const transform: Transform = async (file, api, _options) => {
+  if (!file.source.includes("graphql`")) {
+    return
+  }
+  const j = api.jscodeshift
+  const collection = j(file.source)
+  const schema = await promisedSchema
+
+  collection
+    .find(TaggedTemplateExpression, node => {
+      const tag = node.tag
+      return Identifier.check(tag) && tag.name === "graphql"
+    })
+    .forEach(path => {
+      const templateElement = path.node.quasi.quasis[0]
+      const graphqlDoc = parseGraphQL(templateElement.value.raw)
+
+      const newGraphQLDoc = visit(graphqlDoc, {
+        enter(node, key, parent, path, ancestors) {
+          // let p = Array.isArray(parent) ? parent[0] : parent
+          if (
+            node.kind === "Name" &&
+            parent &&
+            parent.kind == "Field" &&
+            (node.value === "id" ||
+              node.value === "_id" ||
+              node.value === "__id")
+          ) {
+            if (node.value === "__id") {
+              node.value = "id"
+              return node
+            }
+
+            if (node.value === "_id") {
+              node.value = "internalID"
+              return node
+            }
+
+            // Oh boy, it's an `id`, here we go...
+
+            let fragment = ancestors[0]
+            if (fragment && fragment.kind === "FragmentDefinition") {
+              fragment = fragment as FragmentDefinitionNode
+              const fragmentTypeName = fragment.typeCondition.name.value
+              const fragmentType = schema.getType(fragmentTypeName)
+              if (fragmentType) {
+                let parentType: ObjectTypeDefinitionNode
+
+                try {
+                  // Find the parent type of this field selection from the schema...
+                  parentType = walkSelections(
+                    path.slice(2),
+                    fragment,
+                    schema.getType(fragmentTypeName)!.astNode
+                  )
```

Comment:
I think you may be re-implementing some stuff here that comes with graphql-js, namely `visitWithTypeInfo`. E.g.

```ts
const typeInfo = new TypeInfo(schema)

const newGraphQLDoc = visit(
  graphqlDoc,
  visitWithTypeInfo(typeInfo, {
    enter(node, key, parent, path, ancestors) {
      // ...
      typeInfo.getParentType()
    }
  },
)
```

Reply:
Good to know!

Reply:
Thanks for this tip. It _greatly_ simplified everything.

Reply:
💖